### PR TITLE
Add support for multiple CAs

### PR DESCRIPTION
**Pull request by reynir**
The `ca` option supports multiple certificates and even CRLs. Adding support for multiple CA certificates is relatively straight forward.

I'm not sure how much CRLs are used in the wild. The documentation doesn't say CRLs are allowed in the ca file, and from reading the man page on `--capath` it seems it's not supported when openvpn is compiled with mbedtls.

https://github.com/OpenVPN/openvpn/blob/master/src/openvpn/ssl_openssl.c#L1557-L1704

### CODE REVIEW

**Comment by reynir** [comment]
I think this is ready for review. Apparently GitHub defaults to draft pull requests if you made a draft PR previously.

File: src/config.ml [hunk 1]
```diff
@@ -625,11 +635,19 @@ let a_x509_cert_payload ctx constructor str =
   | Error (`Msg msg) -> Error (Fmt.str "%s: invalid certificate: %s" ctx msg)
 
 let a_ca = a_option_with_single_path "ca" `Ca
-let a_ca_payload str = a_x509_cert_payload "CA" (fun c -> B (Ca, c)) str
+let a_ca_payload str =
+  Log.debug (fun m -> m "x509 cert: CA");
```
Comment (hannesm) [inline src/config.ml:639]:
```suggestion
```

File: src/config.ml [hunk 2]
```diff
@@ -625,11 +635,19 @@ let a_x509_cert_payload ctx constructor str =
...
+  match X509.Certificate.decode_pem_multiple (Cstruct.of_string str) with
+  | Ok certs -> Ok (B (Ca, certs))
+  | Error (`Msg msg) -> Error (Fmt.str "ca: invalid certificate(s): %s" msg)
+
 let a_cert = a_option_with_single_path "cert" `Tls_cert
 
 let a_cert_payload str =
-  a_x509_cert_payload "cert" (fun c -> B (Tls_cert, c)) str
+  Log.debug (fun m -> m "x509 cert: cert");
```
Comment (hannesm) [inline src/config.ml:647]:
```suggestion
```

File: src/config.ml [hunk 3]
```diff
@@ -625,11 +635,19 @@ let a_x509_cert_payload ctx constructor str =
   | Error (`Msg msg) -> Error (Fmt.str "%s: invalid certificate: %s" ctx msg)
```
Comment (hannesm) [inline src/config.ml:635]: the `a_x509_cert_payload` is now unused!?

**Comment by hannesm** [comment]
is it possible that we get a test where multiple CA certificates are used?

**Comment by reynir** [comment]
Ah yes, thanks. I have a config and ca file to test this that I apparently didn't commit.

File: src/config.ml (hunk 3 above)
Comment (reynir) [reply src/config.ml:635]: Yes, it was no longer possible to share the code. I removed `a_x509_cert_payload` in a commit I just pushed.

**Comment by hannesm** [comment]
excellent, good to merge once CI is happy

**Comment by hannesm** [comment]
thanks, I squash-merged this. OCaml-CI seems to be unhappy "no space left on device", which is unrelated to this PR.

**Pull request merged**